source/base/hdf5.cc

```diff
@@ -0,0 +1,1563 @@
+    (void)ret;
+  }
+
+
+
+  template <typename Container>
+  void
+  DataSet::write(const Container &data)
+  {
+    AssertDimension(size, internal::get_container_size(data));
+    const std::shared_ptr<hid_t> t_type =
+      internal::get_hdf5_datatype<typename Container::value_type>();
+    hid_t  plist;
+    herr_t ret;
+
+    internal::set_plist(plist, mpi);
+
+    ret = H5Dwrite(*hdf5_reference,
+                   *t_type,
+                   H5S_ALL,
+                   H5S_ALL,
+                   plist,
+                   make_array_view(data).data());
+    Assert(ret >= 0, ExcMessage("Error at H5Dwrite"));
+
+    internal::release_plist(plist,
+                            io_mode,
+                            local_no_collective_cause,
+                            global_no_collective_cause,
+                            mpi,
+                            query_io_mode);
+
+    (void)ret;
+  }
+
+
+
+  template <typename Container>
+  void
+  DataSet::write_selection(const Container &           data,
+                           const std::vector<hsize_t> &coordinates)
+  {
+    AssertDimension(coordinates.size(), data.size() * rank);
+    const std::shared_ptr<hid_t> t_type =
+      internal::get_hdf5_datatype<typename Container::value_type>();
+    const std::vector<hsize_t> data_dimensions =
+      internal::get_container_dimensions(data);
+
+    hid_t  memory_dataspace;
+    hid_t  plist;
+    herr_t ret;
+
+
+    memory_dataspace = H5Screate_simple(1, data_dimensions.data(), NULL);
+    Assert(memory_dataspace >= 0, ExcMessage("Error at H5Screate_simple"));
+    ret = H5Sselect_elements(*dataspace,
+                             H5S_SELECT_SET,
+                             data.size(),
+                             coordinates.data());
+    Assert(ret >= 0, ExcMessage("Error at H5Sselect_elements"));
+
+    internal::set_plist(plist, mpi);
+
+    ret = H5Dwrite(*hdf5_reference,
+                   *t_type,
+                   memory_dataspace,
+                   *dataspace,
+                   plist,
+                   make_array_view(data).data());
+    Assert(ret >= 0, ExcMessage("Error at H5Dwrite"));
+
+    internal::release_plist(plist,
+                            io_mode,
+                            local_no_collective_cause,
+                            global_no_collective_cause,
+                            mpi,
+                            query_io_mode);
+
+    ret = H5Sclose(memory_dataspace);
+    Assert(ret >= 0, ExcMessage("Error at H5SClose"));
+
+    (void)ret;
+  }
+
+
+
+  template <typename Container>
+  void
+  DataSet::write_hyperslab(const Container &           data,
+                           const std::vector<hsize_t> &offset,
+                           const std::vector<hsize_t> &count)
+  {
+    AssertDimension(std::accumulate(count.begin(),
+                                    count.end(),
+                                    1,
+                                    std::multiplies<unsigned int>()),
```

Comment:
In case it helps, below you can find some of the errors given by clang
```c++
dealii/source/base/hdf5.cc:997:7: error: lambda expression in an unevaluated operand
      [&count] {                                                                                            
      ^  
dealii/source/base/hdf5.cc:997:7: error: lambda-expression in unevaluated context
       [&count] {                                                                      
       ^                                                                             
dealii/include/deal.II/base/exceptions.h:1409:32: note: in definition of macro ‘Assert’
         if (__builtin_expect(!(cond), false))                            \                      
                                ^~~~                                                                        
dealii/source/base/hdf5.cc:996:5: note: in expansion of macro ‘AssertDimension’
     AssertDimension(                                                                            
     ^                                                                                                   
dealii/include/deal.II/base/exceptions.h:1570:54: error: template argument 1 is invalid
                                        decltype(dim2)>::type)>::type>(dim1) == \               
                                                      ^                                               
dealii/include/deal.II/base/exceptions.h:1409:32: note: in definition of macro ‘Assert’
         if (__builtin_expect(!(cond), false))                            \                  
                                ^~~~ 
```

Reply:
If you think that it is a bug in deal.ii. I can rebase at several points in time of deal.ii to find out when the lambda function does not compile anymore.

Reply:
This looks similar to https://github.com/nlohmann/json/issues/705. Did you change the compiler in between?

Reply:
It is not the compiler because the code compiles and passes the tests with old versions of deal.ii. I'm taking a systematic approach and rebasing at different points of time of the history of deal.ii. It will take a while, but I think that I can find out the problem :) I'll keep you posted!

Reply:
In the end, I would appreciate if we either find the problem or use `std::accumulate` from the beginning. I am trying to do the latter (originally only to see how much work that would be) and am halfway through. I will push a branch somewhere then.

Reply:
Thanks for your rebase :)
Now the whole history is rebased and std::accumulate is used instead of the lambda function. At each point of the rebase I compiled the code and run the tests to make sure that there were no broken commits.

Reply:
I think that I found the problem for the lambda function. I submitted the issue #7579

Reply:
> At each point of the rebase I compiled the code and run the tests to make sure that there were no broken commits.

Me, too. :wink: